File: src/mbio/packages/bacgenome/complete_assemble.py
# -*- coding: utf-8 -*-
import pandas as pd
import argparse
from Bio import SeqIO
import re,os
import shutil

parser = argparse.ArgumentParser()
parser.add_argument('-f', metavar='[fasta]', required=True, help='fasta file')
parser.add_argument('-d', metavar='[plasmid detail]', required=True, help='plasmid detail file')
parser.add_argument('-p', metavar='[prefix]', required=True, help='prefix of file')
parser.add_argument('-dir', metavar='[dir]', required=True, help='output dir')
args = parser.parse_args()

fasta = args.f
detail = args.d
prefix = args.p
dir = args.dir

def run():
    dict ={}
    if os.path.exists(dir +"/seq_dir"):
        shutil.rmtree(dir +"/seq_dir")
    os.mkdir(dir +"/seq_dir")
    with open(detail) as f:
        lines = f.readlines()
        for line in lines[1:]:
            lin = line.strip().split("\t")
            dict[lin[5]] = lin[4]
    with open(prefix+"_assembly_details.xls","w") as f,open(prefix+"_assembly_summary.xls","w") as k, open("plasmid.type.xls","w") as s:
        k.write("Chromosome No.\tPlasmid No.\tGenome Size (bp)\tG+C (%)\n")
        chrr = []
        plsmid = []
        total_len = 0
        gc_total = 0
        for iterator in SeqIO.parse(fasta, "fasta"):
            SeqIO.write(iterator, dir +"/seq_dir/"+iterator.id + ".fasta", "fasta")
            if re.search("Plasmid", iterator.id) or re.search("plasmid", iterator.id):
                plsmid.append(iterator)
                gene = iterator.id[0].lower()+iterator.id[-1]+"_gene"
                print(iterator.id,gene)
                s.write("{}\t{}\n".format(iterator.id, gene))
            elif re.search("Chromosome", iterator.id) or re.search("chromosome", iterator.id):
                chrr.append(iterator)
            a = iterator.seq.upper().count("A")
            t = iterator.seq.upper().count("T")
            c = iterator.seq.upper().count("C")
            g = iterator.seq.upper().count("G")
            hh = g + c
            gc_total +=hh
            total = len(iterator.seq)
            total_len +=total
            gc = float(g+c)/total*100
            f.write("{}\t{}\t{}\t{}\n".format(iterator.id, total, gc, "A:"+str(a)+";T:"+str(t)+";G:"+str(g)+";C:"+str(c)))
        print(len(chrr))
        if len(chrr) >= 1:
            SeqIO.write(chrr, "chromosome.fasta", "fasta")
        if len(plsmid) >= 1:
            SeqIO.write(plsmid, "plasmid.fasta", "fasta")
        gc = float(gc_total)/total_len*100
        dd =len(chrr)
        ff = len(plsmid)
        k.write("{}\t{}\t{}\t{}\n".format(dd, ff, total_len, gc))

if __name__ == '__main__':
    run()

File: src/mbio/packages/bacgenome/complete_assemble.pl
#! perl -W

use Bio::SeqIO;

my ($list,$dir)=@ARGV;

my @files =glob "$dir/*.fasta";
my %hash;
my %seq;

for my $de (@files){
    my @name=split /\//,$de;
    my $name =$name[-1];
    my $in = Bio::SeqIO->new(-file => "$de", -format => "fasta");
     my $len;
     my $seqs;
     while(my $seq = $in->next_seq())  {
      $len =$seq->length;
      $seqs =$seq->seq;
}
  $hash{$name}=$len;
  $seq{$name}=$seqs;
}
open (IN,$list) || die $!;
my %hash2;
<IN>;
while(<IN>){
  chomp;
  s/[\r\n]//g;
  my @temp =split /\t/;
my $de =@temp;
if($de ==3){
   $hash2{$temp[1]}=$temp[2];
}
}
close IN;

my (@chr,@pla);
foreach my $key (sort {$hash{$b} <=> $hash{$a}} keys %hash){
      if(exists $hash2{$key} && ($hash2{$key} eq 'chromosome' || $hash2{$key} eq 'Chromosome')){
         $des =$key . "\t" . $hash2{$key};
         push @chr,$des;

}elsif(exists $hash2{$key} && ($hash2{$key} eq 'plasmid' || $hash2{$key} eq 'Plasmid')){
           $des =$key . "\t" . $hash2{$key};
          push @pla,$des;
}
}

my %chr=&get_chr(@chr);
my %pla=&get_pla(@pla);
open (OUT,">chromosome.fasta") || dir $!;
foreach (sort keys %chr){
   if(exists $seq{$_}){
    my @temp=split /\t/,$chr{$_};
     my $de=ucfirst($temp[0]);
   print OUT ">$de\n$seq{$_}\n"
}
}
close OUT;
open (OUT2,">plasmid.fasta") || dir $!;
open (OUT3,">plasmid.type.xls") || dir $!;
foreach (sort keys %pla){
   print OUT3 "$pla{$_}\n";
   if(exists $seq{$_}){
    my @temp=split /\t/,$pla{$_};
    my $de=ucfirst($temp[0]);
   print OUT2 ">$de\n$seq{$_}\n";
}
}
close OUT2;
sub get_chr{
   my (@arry)=@_;
   my %new_chr; 
   my $de =@arry;
   if($de == 1){
   my @temp =split /\t/,$arry[0];
   $new_chr{$temp[0]} = $temp[1];
    }else{
    for(my $i=0;$i<$de;$i++){
    my @temp =split /\t/,$arry[$i];
     my $num =$i+1; 
   $new_chr{$temp[0]} = $temp[1] . $num;
    }
   }
 return %new_chr;
}

sub get_pla{
   my (@arry)=@_;
   my %hash =(1=>'A',2 =>'B',3 =>'C',4 =>'D',5 =>'E',6 =>'F',7 =>'G',8 =>'H',9 =>'I',10 =>'J', 11 =>'K', 12 =>'L', 13 =>'M',14 =>'N', 15 =>'O');
   my %new_pla; 
   my $de =@arry;
   if($de == 1){
   my @temp =split /\t/,$arry[0];
   my $de = ucfirst($temp[1]);
   $new_pla{$temp[0]} = $de . "\tp_gene";
}elsif($de != 1){
  for(my $i=0;$i< $de;$i++){
   my @temp =split /\t/,$arry[$i];
   my $de = ucfirst($temp[1]);
   $new_pla{$temp[0]}=$de . $hash{$i+1} . "\t" . "p" . $hash{$i+1} . '_gene';
}
}
return %new_pla; 
}

`cat chromosome.fasta plasmid.fasta >all.fasta`;
